helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java

```diff
@@ -0,0 +1,253 @@
+    // create() semantic will fail if it already exists
+    try {
+      _zkClient.createPersistent(MetadataStoreRoutingConstants.ROUTING_DATA_PATH);
+    } catch (ZkNodeExistsException e) {
+      // This is okay
+    }
+
+    // Get the hostname (REST endpoint) from System property
+    // TODO: Fill in when Helix REST implementations are ready
+    ZNRecord myServerInfo = new ZNRecord("dummy hostname");
+    _leaderElection = new ZkDistributedLeaderElection(_zkClient,
+        MetadataStoreRoutingConstants.LEADER_ELECTION_ZNODE, myServerInfo);
+  }
+
+  @Override
+  public synchronized boolean addMetadataStoreRealm(String realm) {
+    if (_leaderElection.isLeader()) {
+      if (_zkClient.isClosed()) {
+        throw new IllegalStateException("ZkClient is closed!");
+      }
+      return createZkRealm(realm);
+    }
+
+    // TODO: Forward the request to leader
+    return true;
+  }
+
+  @Override
+  public synchronized boolean deleteMetadataStoreRealm(String realm) {
+    if (_leaderElection.isLeader()) {
+      if (_zkClient.isClosed()) {
+        throw new IllegalStateException("ZkClient is closed!");
+      }
+      return _zkClient.delete(MetadataStoreRoutingConstants.ROUTING_DATA_PATH + "/" + realm);
+    }
+
+    // TODO: Forward the request to leader
+    return true;
+  }
+
+  @Override
+  public synchronized boolean addShardingKey(String realm, String shardingKey) {
+    if (_leaderElection.isLeader()) {
+      if (_zkClient.isClosed()) {
+        throw new IllegalStateException("ZkClient is closed!");
+      }
+      // If the realm does not exist already, then create the realm
+      String realmPath = MetadataStoreRoutingConstants.ROUTING_DATA_PATH + "/" + realm;
+      if (!_zkClient.exists(realmPath)) {
+        // Create the realm
+        if (!createZkRealm(realm)) {
+          // Failed to create the realm - log and return false
+          LOG.error(
+              "Failed to add sharding key because ZkRealm creation failed! Namespace: {}, Realm: {}, Sharding key: {}",
+              _namespace, realm, shardingKey);
+          return false;
+        }
+      }
+
+      // Add the sharding key to an empty ZNRecord
+      ZNRecord znRecord;
+      try {
+        znRecord = _zkClient.readData(realmPath);
+      } catch (Exception e) {
+        LOG.error(
+            "Failed to read the realm ZNRecord in addShardingKey()! Namespace: {}, Realm: {}, ShardingKey: {}",
+            _namespace, realm, shardingKey, e);
+        return false;
+      }
+      znRecord.setListField(MetadataStoreRoutingConstants.ZNRECORD_LIST_FIELD_KEY,
+          Collections.singletonList(shardingKey));
+      try {
+        _zkClient.writeData(realmPath, znRecord);
+      } catch (Exception e) {
+        LOG.error(
+            "Failed to write the realm ZNRecord in addShardingKey()! Namespace: {}, Realm: {}, ShardingKey: {}",
+            _namespace, realm, shardingKey, e);
+        return false;
+      }
```

Comment:
This was considered, but raw zookeeper is not very clear about the kind of exceptions it throws (it throws Exception).
Note that as soon as we encounter an exception, we need to return and should not proceed. Otherwise, we might risk getting an NPE.

Reply:
Then let's have a todo here. Since we are working the ZKClient module. Let's have clear different exceptions for different errors. It will benefit the coding convention with that.